src/other/index.jl

```diff
@@ -110,17 +110,31 @@ end
 
 Base.getindex(x::Index, idx::Symbol) = x.lookup[idx]
 Base.getindex(x::AbstractIndex, idx::Real) = Int(idx)
-Base.getindex(x::AbstractIndex, idx::AbstractVector{Union{Bool, Missing}}) =
-    getindex(x, collect(Missings.replace(idx, false)))
+Base.getindex(x::AbstractIndex, idx::AbstractRange) = [idx;]
+Base.getindex(x::AbstractIndex, idx::AbstractVector{<:Real}) = Vector{Int}(idx)
+Base.getindex(x::AbstractIndex, idx::AbstractVector{Symbol}) = [x.lookup[i] for i in idx]
+
 function Base.getindex(x::AbstractIndex, idx::AbstractVector{Bool})
     length(x) == length(idx) || throw(BoundsError(x, idx))
     find(idx)
 end
-Base.getindex(x::AbstractIndex, idx::AbstractVector{T}) where {T >: Missing} =
-    getindex(x, collect(skipmissing(idx)))
-Base.getindex(x::AbstractIndex, idx::AbstractRange) = [idx;]
-Base.getindex(x::AbstractIndex, idx::AbstractVector{T}) where {T <: Real} = convert(Vector{Int}, idx)
-Base.getindex(x::AbstractIndex, idx::AbstractVector{Symbol}) = [x.lookup[i] for i in idx]
+
+# catch all method handling cases when type of idx is not narrowest possible, Any in particular
+# also it handles passing missing values in idx
+function Base.getindex(x::AbstractIndex, idx::AbstractVector)
+    idxs = filter(!ismissing, idx)
+    if length(idxs) != length(idx)
+        Base.depwarn("using missing in column indexing is deprecated", :getindex)
+    end
+    length(idxs) == 0 && return Int[] # special case of empty idxs
+    # below is not very efficient, but has to be done as in Bool case we replaced missing by false
+    # after deprecation period when Missing is disallowed this will not be a problem
+    idxs[1] isa Bool && return getindex(x, Bool[ismissing(v) ? false : v for v in idx])
```

Comment:
I think we'd better handle logical indexing only in the `AbstractVector{Bool}`/`AbstractVector{<: Union{Bool, Missing}}` method. That's what is done for `Array` in Base. No need to make this more complex.

Reply:
Agreed - in 0.7 indexing with `true` is deprecated (in 0.6 it was not). I will remove this support.

A minor question is why do we convert any `Real` to `Int` while Base allows only `Integer`?

Reply:
I guess that's historical. We could deprecate it.